File: windows/AI/Data/ApiTracer.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace AI.Data;

public class TraceRequest
{
    [JsonPropertyName("url")]
    public string Url { get; set; } = "";

    [JsonPropertyName("method")]
    public string Method { get; set; } = "";

    [JsonPropertyName("headers")]
    public Dictionary<string, string> Headers { get; set; } = new();

    [JsonPropertyName("body")]
    public string? Body { get; set; }
}

public class TraceResponse
{
    [JsonPropertyName("statusCode")]
    public int StatusCode { get; set; }

    [JsonPropertyName("headers")]
    public Dictionary<string, string> Headers { get; set; } = new();

    [JsonPropertyName("body")]
    public string? Body { get; set; }
}

public class ApiTrace
{
    [JsonPropertyName("timestamp")]
    public long Timestamp { get; set; }

    [JsonPropertyName("hostname")]
    public string Hostname { get; set; } = "";

    [JsonPropertyName("reportId")]
    public string? ReportId { get; set; }

    [JsonPropertyName("model")]
    public string? Model { get; set; }

    [JsonPropertyName("request")]
    public TraceRequest Request { get; set; } = new();

    [JsonPropertyName("response")]
    public TraceResponse Response { get; set; } = new();

    [JsonIgnore]
    public DateTimeOffset TimestampDate => DateTimeOffset.FromUnixTimeMilliseconds(Timestamp);
}

public class TraceFileInfo
{
    public string Filename { get; set; } = "";
    public string Hostname { get; set; } = "";
    public DateTimeOffset Timestamp { get; set; }
    public int StatusCode { get; set; }
    public string? ReportId { get; set; }
    public string? Model { get; set; }
}

public class ApiTracer
{
    public static readonly ApiTracer Instance = new();

    private static readonly string TraceDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "AI", "trace");

    private readonly object _lock = new();
    private long _fileSequence;
    private static readonly JsonSerializerOptions WriteOptions = new() { WriteIndented = true };

    public bool IsTracingEnabled { get; set; }
    public string? CurrentReportId { get; set; }

    private ApiTracer()
    {
        Directory.CreateDirectory(TraceDir);
    }

    public void TraceIfEnabled(
        string url, string method,
        Dictionary<string, string> requestHeaders, string? requestBody,
        int statusCode, Dictionary<string, string> responseHeaders, string? responseBody,
        DateTimeOffset timestamp)
    {
        if (!IsTracingEnabled) return;

        var hostname = "unknown";
        try { hostname = new Uri(url).Host; } catch { }

        string? model = null;
        if (requestBody != null)
        {
            try
            {
                using var doc = JsonDocument.Parse(requestBody);
                if (doc.RootElement.TryGetProperty("model", out var m))
                    model = m.GetString();
            }
            catch { }
        }

        var trace = new ApiTrace
        {
            Timestamp = timestamp.ToUnixTimeMilliseconds(),
            Hostname = hostname,
            ReportId = CurrentReportId,
            Model = model,
            Request = new TraceRequest { Url = url, Method = method, Headers = requestHeaders, Body = requestBody },
            Response = new TraceResponse { StatusCode = statusCode, Headers = responseHeaders, Body = responseBody }
        };

        SaveTrace(trace);
    }

    public Task TraceIfEnabledAsync(
        string url, string method,
        Dictionary<string, string> requestHeaders, string? requestBody,
        int statusCode, Dictionary<string, string> responseHeaders, string? responseBody,
        DateTimeOffset timestamp)
    {
        TraceIfEnabled(url, method, requestHeaders, requestBody, statusCode, responseHeaders, responseBody, timestamp);
        return Task.CompletedTask;
    }

    private void SaveTrace(ApiTrace trace)
    {
        lock (_lock)
        {
            try
            {
                _fileSequence++;
                var ts = trace.TimestampDate.ToString("yyyyMMdd_HHmmss_fff");
                var seq = Convert.ToString(_fileSequence, 36);
                var filename = $"{trace.Hostname}_{ts}_{seq}.json";
                var json = JsonSerializer.Serialize(trace, WriteOptions);
                File.WriteAllText(Path.Combine(TraceDir, filename), json);
            }
            catch { /* ignore */ }
        }
    }

    public List<TraceFileInfo> GetTraceFiles()
    {
        lock (_lock)
        {
            try
            {
                return Directory.GetFiles(TraceDir, "*.json")
                    .Select(f =>
                    {
                        try
                        {
                            var json = File.ReadAllText(f);
                            var trace = JsonSerializer.Deserialize<ApiTrace>(json);
                            if (trace == null) return null;
                            return new TraceFileInfo
                            {
                                Filename = Path.GetFileName(f),
                                Hostname = trace.Hostname,
                                Timestamp = trace.TimestampDate,
                                StatusCode = trace.Response.StatusCode,
                                ReportId = trace.ReportId,
                                Model = trace.Model
                            };
                        }
                        catch { return null; }
                    })
                    .Where(t => t != null)
                    .OrderByDescending(t => t!.Timestamp)
                    .Cast<TraceFileInfo>()
                    .ToList();
            }
            catch { return new(); }
        }
    }

    public ApiTrace? ReadTrace(string filename)
    {
        lock (_lock)
        {
            try
            {
                var path = Path.Combine(TraceDir, filename);
                if (!File.Exists(path)) return null;
                return JsonSerializer.Deserialize<ApiTrace>(File.ReadAllText(path));
            }
            catch { return null; }
        }
    }

    public void SetTracingEnabled(bool enabled) => IsTracingEnabled = enabled;
    public void SetCurrentReportId(string? id) => CurrentReportId = id;

    public void ClearTraces()
    {
        lock (_lock)
        {
            try
            {
                foreach (var f in Directory.GetFiles(TraceDir, "*.json"))
                    File.Delete(f);
            }
            catch { /* ignore */ }
        }
    }

    public int GetTraceCount()
    {
        lock (_lock)
        {
            try { return Directory.GetFiles(TraceDir, "*.json").Length; }
            catch { return 0; }
        }
    }
}

File: macOS/macAI/Data/ApiTracer.swift
import Foundation
import os

// MARK: - Trace Models

struct TraceRequest: Codable {
    let url: String
    let method: String
    let headers: [String: String]
    let body: String?
}

struct TraceResponse: Codable {
    let statusCode: Int
    let headers: [String: String]
    let body: String?
}

struct ApiTrace: Codable {
    let timestamp: Date
    let hostname: String
    let reportId: String?
    let model: String?
    let request: TraceRequest
    let response: TraceResponse
}

struct TraceFileInfo: Identifiable {
    let id = UUID()
    let filename: String
    let hostname: String
    let timestamp: Date
    let statusCode: Int
    let reportId: String?
    let model: String?
}

// MARK: - API Tracer Actor

/// Thread-safe API trace storage using actor.
actor ApiTracer {
    static let shared = ApiTracer()

    private static let logger = Logger(subsystem: "com.ai.macAI", category: "ApiTracer")
    private let traceDir: URL
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder
    private var fileSequence: Int64 = 0

    var isTracingEnabled = false
    var currentReportId: String?

    private init() {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        traceDir = appSupport.appendingPathComponent("com.ai.macAI/trace", isDirectory: true)
        try? FileManager.default.createDirectory(at: traceDir, withIntermediateDirectories: true)

        encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .millisecondsSince1970

        decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .millisecondsSince1970
    }

    // MARK: - Trace If Enabled

    func traceIfEnabled(
        url: String,
        method: String,
        requestHeaders: [String: String],
        requestBody: String?,
        statusCode: Int,
        responseHeaders: [String: String],
        responseBody: String?,
        timestamp: Date
    ) {
        guard isTracingEnabled else { return }

        let hostname = URL(string: url)?.host ?? "unknown"

        // Extract model from request body
        var model: String?
        if let body = requestBody, let data = body.data(using: .utf8) {
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                model = json["model"] as? String
            }
        }

        let trace = ApiTrace(
            timestamp: timestamp,
            hostname: hostname,
            reportId: currentReportId,
            model: model,
            request: TraceRequest(url: url, method: method, headers: requestHeaders, body: requestBody),
            response: TraceResponse(statusCode: statusCode, headers: responseHeaders, body: responseBody)
        )

        saveTrace(trace)
    }

    // MARK: - Save

    private func saveTrace(_ trace: ApiTrace) {
        fileSequence += 1
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmmss_SSS"
        let timestamp = formatter.string(from: trace.timestamp)
        let seq = String(fileSequence, radix: 36)
        let filename = "\(trace.hostname)_\(timestamp)_\(seq).json"
        let fileURL = traceDir.appendingPathComponent(filename)

        do {
            let data = try encoder.encode(trace)
            try data.write(to: fileURL)
        } catch {
            Self.logger.error("Failed to save trace: \(error.localizedDescription)")
        }
    }

    // MARK: - Read

    func getTraceFiles() -> [TraceFileInfo] {
        do {
            let files = try FileManager.default.contentsOfDirectory(at: traceDir, includingPropertiesForKeys: nil)
                .filter { $0.pathExtension == "json" }

            return files.compactMap { fileURL -> TraceFileInfo? in
                guard let data = try? Data(contentsOf: fileURL),
                      let trace = try? decoder.decode(ApiTrace.self, from: data) else { return nil }
                return TraceFileInfo(
                    filename: fileURL.lastPathComponent,
                    hostname: trace.hostname,
                    timestamp: trace.timestamp,
                    statusCode: trace.response.statusCode,
                    reportId: trace.reportId,
                    model: trace.model
                )
            }.sorted { $0.timestamp > $1.timestamp }
        } catch {
            return []
        }
    }

    func readTrace(_ filename: String) -> ApiTrace? {
        let fileURL = traceDir.appendingPathComponent(filename)
        guard let data = try? Data(contentsOf: fileURL) else { return nil }
        return try? decoder.decode(ApiTrace.self, from: data)
    }

    func setTracingEnabled(_ enabled: Bool) {
        isTracingEnabled = enabled
    }

    func setCurrentReportId(_ id: String?) {
        currentReportId = id
    }

    func clearTraces() {
        do {
            let files = try FileManager.default.contentsOfDirectory(at: traceDir, includingPropertiesForKeys: nil)
            for file in files where file.pathExtension == "json" {
                try FileManager.default.removeItem(at: file)
            }
        } catch {
            Self.logger.error("Failed to clear traces: \(error.localizedDescription)")
        }
    }

    func getTraceCount() -> Int {
        (try? FileManager.default.contentsOfDirectory(at: traceDir, includingPropertiesForKeys: nil)
            .filter { $0.pathExtension == "json" }.count) ?? 0
    }
}
